### PR TITLE
Add best practices for HTTP 301 and retries to Spark vignette

### PR DESCRIPTION
**Pull request by mweiden**
When querying the DSS API, clients are expected
* Handle HTTP 301 responses and retry their requests after the number of seconds specified in the `retry-after` header
* Retry their requests to the API as requests can fail when the api is under high load

This PR demonstrates these best practices in the data consumer vignettes.

Fixes https://github.com/HumanCellAtlas/data-store/issues/1263

### CODE REVIEW

File: tasks/SPARK/src/main/java/kmer/Kmer.java [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+ * N: to find top-N
+ * <p>
+ * A kmer or k-mer is a short DNA sequence consisting of a fixed
+ * number (K) of bases. The value of k is usually divisible by 4
+ * so that a kmer can fit compactly into a basevector object.
+ * Typical values include 12, 20, 24, 36, and 48; kmers of these
+ * sizes are referred to as 12-mers, 20-mers, and so forth.
+ *
+ * @author Mahmoud Parsian
+ * @author Brian O'Connor (made tweaks for HCA demo)
```
Comment (briandoconnor) [inline tasks/SPARK/src/main/java/kmer/Kmer.java]: @mweiden you should add yourself as an author too, right?

File: tasks/SPARK/src/test/java/kmer/DataStoreTest.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
+package kmer;
```
Comment (briandoconnor) [inline tasks/SPARK/src/test/java/kmer/DataStoreTest.java]: So happy to see you added tests!!!

File: tasks/SPARK/src/main/java/kmer/HttpUtil.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+            String jsonText = readAll(rd);
+            JSONObject json = new JSONObject(jsonText);
+            return json;
+        } finally {
+            is.close();
+        }
+    }
+
+    public static JSONObject jsonRequest(String url) throws IOException, InterruptedException {
+        Optional<JSONObject> maybeJSON = Retriable.runWithRetries(
```
Comment (briandoconnor) [inline tasks/SPARK/src/main/java/kmer/HttpUtil.java]: Great!!!

File: tasks/SPARK/src/test/java/kmer/RetriableTest.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+        public int getCounter() {
+            return counter;
+        }
+    }
+
+    @Test
+    public void runWithRetriesShouldRetry() throws InterruptedException {
+        TestRetriable retriable = new TestRetriable();
+        Retriable.runWithRetries( 3, 0, retriable);
```
Comment (briandoconnor) [inline tasks/SPARK/src/test/java/kmer/RetriableTest.java]: Do you want to try a retry test that uses a mock web server to return a 503 (that's what the blue box is using to signal a try again later, right?)

File: tasks/SPARK/src/main/java/kmer/Retriable.java [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+import java.util.Random;
+
+public interface Retriable<T> {
+    T run() throws Exception;
+    Random random = new Random();
+
+    static <T> Optional<T> runWithRetries(int maxRetries, int maxInterval, Retriable<T> t) throws InterruptedException {
+        int count = 0;
+        while (count < maxRetries) {
+            try {
```
Comment (briandoconnor) [inline tasks/SPARK/src/main/java/kmer/Retriable.java]: Does a 503 really cause an exception?  Does this need to be tested through a unit test?

Comment (mweiden) [reply tasks/SPARK/src/main/java/kmer/Retriable.java]: Yep. Very Java.

**Comment by mweiden** [comment]
@briandoconnor I addressed your comments and made some other improvements including command line arg configuration with flags. I've run this successfully against the prod deployment. Merging this now.

**Pull request merged**